content/blog/zephyr-tracing-and-profiling/index.md

```diff
@@ -0,0 +1,279 @@
+﻿---
+title: "Introduction to Zephyr OS Tracing and Profiling"
+date: 2024-10-25T06:54:17+08:00
+showAuthor: false
+featureAsset: "img/featured/featured-espressif.webp"
+authors:
+  - "raffael-rostagno"
+tags: ["ESP32", "Tracing",  "Profiling","Zephyr"]
+---
+
+## Overview
+
+Embedded systems can be quite complex, depending on the architecture, application size and nature. Even for an experienced Embedded developer, understanding the interplay of threads, interrupts and many processes that run in a large application can be challenging in a situation where debugging, evaluating hardware adequacy or optimizing application code is necessary. We humans are (mostly) visual by nature, and having the means to *visualize* what is happening in a given system can really open up possibilities. Without the right tools we are often in the dark, quite literally.
+
+One category of tools interesting to learn about is Tracing and Profiling. *Tracing* means to collect software execution data - the sequence and timing of execution of functions, threads, ISRs - and log it for posterior analysis, without interrupting or interfering (at best) in the application execution. *Profiling* aims to measure performance metrics such as execution time, CPU usage and dynamic memory usage. With such tools it is possible to evaluate how well an embedded system is performing, understand if the hardware used is adequate for the application being run and debug eventual problems.
+
+In this article we are going to do a brief introduction and show some basic examples of these resources as implemented in Zephyr RTOS. Later we’ll take a look at some of the features offered by Percepio’s Tracealyzer, which is a third party solution that extends some of Zephyr's *tracing and profiling* native functionalities and offers a frontend tool to analyze the data.
+
+
+## Zephyr’s tracing tool
+
+In order to show how Zephyr’s native tracing system works, we’ll use the sample below and an ESP32-C6 to run it on a physical device:
+
+`samples/subsys/tracing`
+
+
+To enable Zephyr’s native tracing, a small number of options are available. For this example, which uses ESP32-C6’s USB JTAG port to output the logged data, we’ll set the following:
+
+```
+CONFIG_TRACING=y
+CONFIG_TRACING_CTF=y
+CONFIG_TRACING_BACKEND_UART=y
+CONFIG_TRACING_BUFFER_SIZE=4096
+```
+
+These settings activate the tracing feature, defines the data stored format as CTF (Common Trace Format) and configures the output medium (UART). As UART was selected as the medium to transfer data from the device to the host (PC), we’ll define the following settings in an overlay file within the sample:
+
+```dts
+/ {
+    chosen {
+        zephyr,tracing-uart = &usb_serial;
+    };
+};
+
+&usb_serial {
+    status = "okay";
+};
+```
+
+In order to build the sample we’ll use the following command:
+
+```console

```

Comment:
IIRC the content type should be "shell". Also, I'm not sure if "dts" means something to this renderer.